resources/agent.rb

```diff
@@ -35,32 +35,66 @@
                                    "backend-url": ['ws://127.0.0.1:8081'],
                                  }
 action :install do
-  packagecloud_repo new_resource.repo do
-    type value_for_platform_family(
-      %w(rhel fedora amazon) => 'rpm',
-      'default' => 'deb'
-    )
-  end
+  # Linux installation - source package
+  if node['platform'] != 'windows'
+    packagecloud_repo new_resource.repo do
+      type value_for_platform_family(
+        %w(rhel fedora amazon) => 'rpm',
+        'default' => 'deb'
+      )
+    end
 
-  package 'sensu-go-agent' do
-    if latest_version?(new_resource.version)
-      action :upgrade
-    else
-      version new_resource.version
-      action :install
+    # Installs/upgrades sensu-go-agent package
+    package 'sensu-go-agent' do
+      if latest_version?(new_resource.version)
+        action :upgrade
+      else
+        version new_resource.version
+        action :install
+      end
     end
-  end
 
-  # render template at /etc/sensu/agent.yml
-  file ::File.join(new_resource.config_home, 'agent.yml') do
-    content(stringify_keys(new_resource.config))
+    # render template at /etc/sensu/agent.yml for linux
+    file ::File.join(new_resource.config_home, 'agent.yml') do
+      content(stringify_keys(new_resource.config))
+    end
+
+    # Enable and start the sensu-agent service
+    service 'sensu-agent' do
+      if node['platform'] == 'ubuntu' && node['platform_version'].to_f == 14.04
+        provider Chef::Provider::Service::Init
+        action :start
+      else
+        action [:enable, :start]
+      end
+    end
   end
 
-  service 'sensu-agent' do
-    if node['platform'] == 'ubuntu' && node['platform_version'].to_f == 14.04
-      provider Chef::Provider::Service::Init
-      action :start
-    else
+  # Installs msi for Sensu-Go
+  if node['platform'] == 'windows'
+    windows_package 'sensu-go-agent' do
+      source 'https://s3-us-west-2.amazonaws.com/sensu.io/sensu-go/5.10.0/sensu-go-agent_5.10.0.4171_en-US.x64.msi'
+      installer_type :custom
+      version '5.10.0.4171'
+    end
+
+    # Adds install directory to path
+    windows_path 'c:\\Program Files\\sensu\\sensu-agent\\bin'
+
+    # render template at c:\Programdata\Sensu\config\agent.yml for windows
+    file ::File.join('c:/ProgramData/Sensu/config', 'agent.yml') do
+      content(stringify_keys(new_resource.config))
+    end
+
+    # Installs SensuAgent Service
+    powershell_script 'SensuAgent Service' do
+      code '.\\sensu-agent.exe service install'
+      cwd 'c:/Program Files/sensu/sensu-agent/bin'
+      not_if '((Get-Service SensuAgent).Name -eq "SensuAgent")'
```

Comment:
this would mean if there are changes to the service script (not sure if thats the right phrase for windows) it will never be upgraded. Should we consider having an argument to say please force upgrade? :thinking:

Reply:
Even better would be check the version and only upgrade on new versions.

Reply:
This resource only handles creating the service if it does not already exist, thus the guard.

Reply:
That's exactly my point, if the service exists but some config for the service has changed this will never get updated. We could also have this do an action of `:nothing` and have a notifies to control when its executed to avoid chicken and egg scenarios. That being said I don't know much about windows but I imagine that changes to the service definitions are important and control things like logging, log rotation, etc.

Reply:
Upgrades and downgrades respected the current implementation. I could support this being enhanced if functionality around the binary is increased, but as is, that's the only function provided with `service install`